com.unity.render-pipelines.universal/Samples~/TeachingSamples/SharedAssets/Scripts/FPSController.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+[RequireComponent(typeof(Rigidbody))]
+public class FPSController : MonoBehaviour
+{
+    public bool moveWithMouse = true;
+    public float mouseSensitivity = 1;
+    public float buttonSensitivity = 1;
+    public float walkSpeed = 1;
+
+    private Rigidbody rb;
+    private byte movementFlags;
+
+    // Start is called before the first frame update
+    void Start()
+    {
+        if (moveWithMouse)
+        {
+            Cursor.visible = false;
+            Cursor.lockState = CursorLockMode.Locked;
+        }
+        
+        rb = GetComponent<Rigidbody>();
+
+        movementFlags = 0;
+    }
+
+    private void Update()
+    {
+        if (moveWithMouse)
+        {
+            //Rotate based on mouse input
+            float vert = Input.GetAxis("Mouse X") * Time.deltaTime * mouseSensitivity * 100;
+            float hori = Input.GetAxis("Mouse Y") * Time.deltaTime * mouseSensitivity * 100;
+            transform.Rotate(Vector3.up * vert, Space.World);
+            transform.Rotate(-Vector3.right * hori, Space.Self);
+        }
+
+        Vector3 rotation = MovementMaskToRotation() * (Time.deltaTime * buttonSensitivity * 100);
+        transform.Rotate(0, rotation.y, 0, Space.World);
+        transform.Rotate(rotation.x, 0, 0, Space.Self);
+
+    }
+
+    // Update is called once per frame
+    void FixedUpdate()
+    {
+        //Move based on wasd input
+        Vector3 forward = transform.forward;
+        forward.y = 0;
+        forward = forward.normalized;
+
+        Vector3 direction = Vector3.zero;
+
+
+
+        direction = MovementMaskToWalkDirection();
+        
+
+        if (Input.GetKey(KeyCode.W))
+        {
+            direction += forward;
+        }
+        if (Input.GetKey(KeyCode.S))
+        {
+            direction -= forward;
+        }
+        if (Input.GetKey(KeyCode.A))
+        {
+            direction -= transform.right;
+        }
+        if (Input.GetKey(KeyCode.D))
+        {
+            direction += transform.right;
+        }
+        direction = direction.normalized;
+
+        rb.AddForce((direction * walkSpeed) - rb.velocity, ForceMode.VelocityChange);
```

Comment:
We are taking away any force the controller has from previous frames? Does this take away gravity too? Wouldnt RigidBody.MovePosition() be more suitable for this behavior?

Reply:
I had the impression that RigidBody.MovePosition() was for manipulating kinematic bodies, which the player rigidbody is not.